store/v2/commitment/store_test_suite.go

```diff
@@ -166,3 +221,83 @@ func (s *CommitStoreTestSuite) TestStore_Pruning() {
 		}
 	}
 }
+
+func (s *CommitStoreTestSuite) TestStore_Upgrades() {
+	storeKeys := []string{storeKey1, storeKey2, storeKey3}
+	commitDB := dbm.NewMemDB()
+	commitStore, err := s.NewStore(commitDB, storeKeys, log.NewNopLogger())
+	s.Require().NoError(err)
+
+	latestVersion := uint64(10)
+	kvCount := 10
+	for i := uint64(1); i <= latestVersion; i++ {
+		kvPairs := make(map[string]corestore.KVPairs)
+		for _, storeKey := range storeKeys {
+			kvPairs[storeKey] = corestore.KVPairs{}
+			for j := 0; j < kvCount; j++ {
+				key := []byte(fmt.Sprintf("key-%d-%d", i, j))
+				value := []byte(fmt.Sprintf("value-%d-%d", i, j))
+				kvPairs[storeKey] = append(kvPairs[storeKey], corestore.KVPair{Key: key, Value: value})
+			}
+		}
+		s.Require().NoError(commitStore.WriteChangeset(corestore.NewChangesetWithPairs(kvPairs)))
+		_, err = commitStore.Commit(i)
+		s.Require().NoError(err)
+	}
+
+	// create a new commitment store with upgrades
+	upgrades := &corestore.StoreUpgrades{
+		Added: []string{"newStore1", "newStore2"},
+		Renamed: []corestore.StoreRename{
+			{OldKey: storeKey1, NewKey: "renamedStore1"},
+		},
+		Deleted: []string{storeKey3},
+	}
+	newStoreKeys := []string{"store1", "store2", "store3", "renamedStore1", "newStore1", "newStore2"}
+	realStoreKeys := []string{"renamedStore1", "store2", "newStore1", "newStore2"}
+	commitStore, err = s.NewStore(commitDB, newStoreKeys, log.NewNopLogger())
+	s.Require().NoError(err)
+	err = commitStore.LoadVersionAndUpgrade(latestVersion, upgrades)
+	s.Require().NoError(err)
+
+	// apply the changeset again
+	for i := latestVersion + 1; i < latestVersion*2; i++ {
```

Comment:
At this stage, I would expect read operations on the store only so that you have confidence that everything is ok.
When you write to the store, the base changes.

I would assert
* "newStore1", "newStore2" exists with initial state
* "renamedStore1" exists with state that was in in `storeKey1` before.
* ensure store `storeKey1` does not exists anymore
* ensure store `storeKey3` does not exists anymore
* ensure store `storeKey2` state was not touched

Reply:
I am sorry but I still don't get the `apply the changeset again` step.  IMHO there is no need for this. The state exists and was handled on the upgrade. All you have to do now is verify.
With writing to the store again, you would overwrite the updated data. I have removed L272-291 locally and everything was passing. What do I miss?